middleware/standard_response.go

```diff
@@ -0,0 +1,168 @@
+	}
+
+	StandardResponseParams struct {
+		Handle func(w http.ResponseWriter, r *http.Request, data interface{}, err error)
+	}
+
+	standardResponseCtxKey struct{}
+)
+
+func NewStandardResponseHandler(params StandardResponseHandlerParams) *StandardResponseHandler {
+	return &StandardResponseHandler{
+		params: params,
+	}
+}
+
+func NewStandardResponse(params StandardResponseParams) *StandardResponse {
+	return &StandardResponse{
+		params: params,
+	}
+}
+
+func GetStandardResponseHandler(ctx context.Context) func(interface{}, error) {
+	v := ctx.Value(standardResponseCtxKey{})
+	if v == nil {
+		return nil
+	}
+	return v.(func(data interface{}, err error))
+}
+
+func (h *StandardResponseHandler) GetHandleBody(
+	r *http.Request,
+	data interface{},
+	err error,
+) (httpStatus int, body interface{}) {
+	httpStatus = http.StatusOK
+	bodyType := StandardResponseBodyJson
+	if h.params.CheckBodyType != nil {
+		bodyType = h.params.CheckBodyType(r)
+	}
+
+	if err != nil {
+		e, ok := errorx.AsCodeError(err)
+		if !ok {
+			err = errorx.WithCode(errorx.TakeCodePriority(func() *errorx.ErrCode {
+				if h.params.GetErrCode == nil {
+					return nil
+				}
+				return h.params.GetErrCode(err)
+			}, func() *errorx.ErrCode {
+				return errorx.NewErrCode(errorx.CCInternalServer, 0, 0, "ErrInternalServer")
+			}), err)
+			e, _ = errorx.AsCodeError(err)
+		}
+		httpStatus = e.GetHTTPStatus()
+
+		if bodyType != StandardResponseBodyNone {
+			resp := map[string]interface{}{
+				standardResponseFieldCode:    e.GetCode(),
+				standardResponseFieldMessage: e.GetMessage(),
+			}
+			if h.params.DebugInfo {
+				resp[standardResponseFieldDebugInfo] = fmt.Sprintf("%+v", err)
+			}
+			body = resp
+		}
+	} else if bodyType != StandardResponseBodyNone {
+		resp := map[string]interface{}{
+			standardResponseFieldCode:    0,
+			standardResponseFieldMessage: "Success",
+		}
+		if data != nil {
+			resp[standardResponseFieldData] = data
+		}
+		body = resp
+	}
+
+	return httpStatus, body
+}
+
+func (h *StandardResponseHandler) Handle(
+	w http.ResponseWriter,
+	r *http.Request,
+	data interface{},
+	err error,
+) {
+	httpStatus, body := h.GetHandleBody(r, data, err)
+	if body == nil {
+		w.WriteHeader(httpStatus)
+		return
+	}
+
+	bs, err := json.Marshal(body)
+	if err != nil {
+		h.errorf("write response json.Marshal failed, error: %s", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(httpStatus)
+	if n, err := w.Write(bs); err != nil {
+		if err != http.ErrHandlerTimeout {
+			h.errorf("write response failed, error: %s", err)
+		}
+	} else if n < len(bs) {
+		h.errorf("write response failed, actual bytes: %d, written bytes: %d", len(bs), n)
+	}
+}
+
+func (m *StandardResponse) Handler(next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		next(w, r.WithContext(context.WithValue(r.Context(),
+			standardResponseCtxKey{},
```

Comment:
standardResponseCtxKey no need to be uppercase？

Reply:
This does not need to be exposed to avoid misuse.